### PR TITLE
Consumer groups

### PR DESCRIPTION
**Pull request by andrea-vinci**
- [ ] Show consumer group state in subtitle
- [ ] Show members and associated partitions
- [ ] Allow reset offset *only* if state is Empty

### CODE REVIEW

File: src/main/kotlin/insulator/lib/kafka/model/ConsumerGroup.kt [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package insulator.lib.kafka.model
+
+import org.apache.kafka.common.ConsumerGroupState
+
+data class ConsumerGroup(val id: String, val state: ConsumerGroupState, val members: List<ConsumerGroupMember>)
```
Comment (darka91) [inline src/main/kotlin/insulator/lib/kafka/model/ConsumerGroup.kt:5]: Codacy found an issue: [UndocumentedPublicClass - [ConsumerGroup] at /src/src/main/kotlin/insulator/lib/kafka/model/ConsumerGroup.kt:5:12](https://app.codacy.com/manual/darka91/Insulator/pullRequest?prid=5960458)

**Comment by darka91** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 14
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/darka91/Insulator/pullRequest?prid=5960458&bid=19543099)

File: src/main/kotlin/insulator/viewmodel/main/consumergroup/ConsumerGroupViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+                            }
+                        )
+                    }
+
+                consumerGroupMembers.addAll(sorted)
+                state.value = it.state.toString()
+            }.handleErrorWith { print(it) } // todo: pass to the UI
+    }
+
+    fun resetOffset() {
```
Comment (darka91) [inline src/main/kotlin/insulator/viewmodel/main/consumergroup/ConsumerGroupViewModel.kt:39]: Codacy found an issue: [UndocumentedPublicFunction - [resetOffset] at /src/src/main/kotlin/insulator/viewmodel/main/consumergroup/ConsumerGroupViewModel.kt:39:9](https://app.codacy.com/manual/darka91/Insulator/pullRequest?prid=5960458)

File: src/main/kotlin/insulator/views/main/consumergroup/ListConsumerGroupView.kt [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+import insulator.viewmodel.main.consumergroup.ConsumerGroupViewModel
+import insulator.viewmodel.main.consumergroup.ListConsumerGroupViewModel
+import insulator.views.common.searchBox
+import javafx.beans.property.SimpleStringProperty
+import javafx.scene.control.SelectionMode
+import javafx.scene.layout.Priority
+import tornadofx.* // ktlint-disable no-wildcard-imports
+
+class ListConsumerGroupView : View("Consumer groups") {
```
Comment (darka91) [inline src/main/kotlin/insulator/views/main/consumergroup/ListConsumerGroupView.kt:11]: Codacy found an issue: [UndocumentedPublicClass - [ListConsumerGroupView] at /src/src/main/kotlin/insulator/views/main/consumergroup/ListConsumerGroupView.kt:11:7](https://app.codacy.com/manual/darka91/Insulator/pullRequest?prid=5960458)

File: src/main/kotlin/insulator/views/main/consumergroup/ListConsumerGroupView.kt [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+    private val viewModel: ListConsumerGroupViewModel by inject()
+    private val searchItem = SimpleStringProperty()
+
+    override val root = vbox(spacing = 5.0) {
+        searchBox(searchItem)
+        listview<String> {
+            cellFormat { graphic = label(it) }
+            onDoubleClick {
+                if (this.selectedItem == null) return@onDoubleClick
```
Comment (darka91) [inline src/main/kotlin/insulator/views/main/consumergroup/ListConsumerGroupView.kt:21]: Codacy found an issue: [LabeledExpression - [<anonymous>] at /src/src/main/kotlin/insulator/views/main/consumergroup/ListConsumerGroupView.kt:21:48](https://app.codacy.com/manual/darka91/Insulator/pullRequest?prid=5960458)

File: src/main/kotlin/insulator/lib/helpers/CompletableFutureHelper.kt [hunk 5]
```diff
@@ -25,3 +27,17 @@ fun <A, B> CompletableFuture<Either<Throwable, A>>.flatMap(f: (A) -> Either<Thro
...
+    this.thenApply {
+        Platform.runLater {
+            it.runCatching { map(f) }
+                .fold({ result.complete(Option.empty()) }, { result.complete(Option.just(it)) })
+        }
+    }
+    return result
+}
+
+fun CompletableFuture<Option<Throwable>>.handleErrorWith(f: (Throwable) -> Unit): CompletableFuture<Unit> =
```
Comment (darka91) [inline src/main/kotlin/insulator/lib/helpers/CompletableFutureHelper.kt:42]: Codacy found an issue: [UndocumentedPublicFunction - [handleErrorWith] at /src/src/main/kotlin/insulator/lib/helpers/CompletableFutureHelper.kt:42:42](https://app.codacy.com/manual/darka91/Insulator/pullRequest?prid=5960458)

File: src/main/kotlin/insulator/lib/kafka/AdminApi.kt [hunk 6]
```diff
@@ -29,5 +36,32 @@ class AdminApi(private val admin: AdminClient, private val consumer: Consumer<An
                 }
             }
 
+    fun listConsumerGroups() = admin.listConsumerGroups().all().toCompletableFuture()
```
Comment (darka91) [inline src/main/kotlin/insulator/lib/kafka/AdminApi.kt:39]: Codacy found an issue: [UndocumentedPublicFunction - [listConsumerGroups] at /src/src/main/kotlin/insulator/lib/kafka/AdminApi.kt:39:9](https://app.codacy.com/manual/darka91/Insulator/pullRequest?prid=5960458)

File: src/test/kotlin/insulator/lib/kafka/AdminApiTest.kt [hunk 7]
```diff
@@ -75,4 +83,37 @@ class AdminApiTest : FunSpec({
         // assert
         res.get() shouldBeRight listOf(Topic("topic1", true, 1, 5), Topic("topic2", false, 2, 5))
     }
+
+    test("describeConsumerGroup happy path") {
+        // arrange
+        val topicName = "topic1"
+        val (consumerGroupOffset, partitionEndOffset) = Pair(1L, 5L)
+        val consumerGroupDescription = mockk<ConsumerGroupDescription> {
+            every { groupId() } returns "consumerGroup"
```
Comment (darka91) [inline src/test/kotlin/insulator/lib/kafka/AdminApiTest.kt:92]: Codacy found an issue: [StringLiteralDuplication - 3/3 - [<anonymous>] at /src/src/test/kotlin/insulator/lib/kafka/AdminApiTest.kt:92:41](https://app.codacy.com/manual/darka91/Insulator/pullRequest?prid=5960458)

File: src/main/kotlin/insulator/viewmodel/main/consumergroup/ConsumerGroupViewModel.kt [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+import insulator.di.getInstanceNow
+import insulator.lib.helpers.handleErrorWith
+import insulator.lib.helpers.runOnFXThread
+import insulator.lib.kafka.AdminApi
+import javafx.beans.property.SimpleStringProperty
+import javafx.collections.FXCollections
+import javafx.collections.ObservableList
+import tornadofx.ViewModel
+
+class ConsumerGroupViewModel(consumerGroupId: String) : ViewModel() {
```
Comment (darka91) [inline src/main/kotlin/insulator/viewmodel/main/consumergroup/ConsumerGroupViewModel.kt:12]: Codacy found an issue: [UndocumentedPublicClass - [ConsumerGroupViewModel] at /src/src/main/kotlin/insulator/viewmodel/main/consumergroup/ConsumerGroupViewModel.kt:12:7](https://app.codacy.com/manual/darka91/Insulator/pullRequest?prid=5960458)

File: src/main/kotlin/insulator/lib/kafka/AdminApi.kt [hunk 9]
```diff
@@ -29,5 +36,32 @@ class AdminApi(private val admin: AdminClient, private val consumer: Consumer<An
...
+        .map { consumerGroup -> consumerGroup.map { it.groupId() } }
+
+    fun describeConsumerGroup(groupId: String): CompletableFuture<Either<Throwable, ConsumerGroup>> =
```
Comment (darka91) [inline src/main/kotlin/insulator/lib/kafka/AdminApi.kt:42]: Codacy found an issue: [UndocumentedPublicFunction - [describeConsumerGroup] at /src/src/main/kotlin/insulator/lib/kafka/AdminApi.kt:42:9](https://app.codacy.com/manual/darka91/Insulator/pullRequest?prid=5960458)

File: src/main/kotlin/insulator/views/main/consumergroup/ConsumerGroupView.kt [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+import javafx.beans.binding.Bindings
+import javafx.beans.property.SimpleStringProperty
+import javafx.collections.FXCollections
+import javafx.geometry.Pos
+import javafx.scene.control.TreeItem
+import javafx.scene.layout.Priority
+import tornadofx.* // ktlint-disable no-wildcard-imports
+import java.util.concurrent.Callable
+
+class ConsumerGroupView : View() {
```
Comment (darka91) [inline src/main/kotlin/insulator/views/main/consumergroup/ConsumerGroupView.kt:19]: Codacy found an issue: [UndocumentedPublicClass - [ConsumerGroupView] at /src/src/main/kotlin/insulator/views/main/consumergroup/ConsumerGroupView.kt:19:7](https://app.codacy.com/manual/darka91/Insulator/pullRequest?prid=5960458)

File: src/main/kotlin/insulator/lib/kafka/model/ConsumerGroup.kt [hunk 11]
```diff
@@ -0,0 +1,7 @@
...
+data class ConsumerGroupMember(val clientId: String, val topicPartitions: List<TopicPartitionLag>)
+data class TopicPartitionLag(val topic: String, val partition: Int, val lag: Long)
```
Comment (darka91) [inline src/main/kotlin/insulator/lib/kafka/model/ConsumerGroup.kt:7]: Codacy found an issue: [UndocumentedPublicClass - [TopicPartitionLag] at /src/src/main/kotlin/insulator/lib/kafka/model/ConsumerGroup.kt:7:12](https://app.codacy.com/manual/darka91/Insulator/pullRequest?prid=5960458)

File: src/main/kotlin/insulator/viewmodel/main/consumergroup/ListConsumerGroupViewModel.kt [hunk 12]
```diff
@@ -0,0 +1,20 @@
+package insulator.viewmodel.main.consumergroup
+
+import insulator.lib.helpers.map
+import insulator.lib.kafka.AdminApi
+import javafx.collections.FXCollections
+import javafx.collections.ObservableList
+import tornadofx.ViewModel
+
+class ListConsumerGroupViewModel : ViewModel() {
```
Comment (darka91) [inline src/main/kotlin/insulator/viewmodel/main/consumergroup/ListConsumerGroupViewModel.kt:9]: Codacy found an issue: [UndocumentedPublicClass - [ListConsumerGroupViewModel] at /src/src/main/kotlin/insulator/viewmodel/main/consumergroup/ListConsumerGroupViewModel.kt:9:7](https://app.codacy.com/manual/darka91/Insulator/pullRequest?prid=5960458)

File: src/main/kotlin/insulator/lib/helpers/CompletableFutureHelper.kt [hunk 13]
```diff
@@ -25,3 +27,17 @@ fun <A, B> CompletableFuture<Either<Throwable, A>>.flatMap(f: (A) -> Either<Thro
 
 fun <A, B> CompletableFuture<Either<Throwable, A>>.fold(ifLeft: (Throwable) -> B, ifRight: (A) -> B):
     CompletableFuture<B> = this.thenApply { it.fold(ifLeft, ifRight) }
+
+fun <A> CompletableFuture<Either<Throwable, A>>.runOnFXThread(f: (A) -> Unit): CompletableFuture<Option<Throwable>> {
```
Comment (darka91) [inline src/main/kotlin/insulator/lib/helpers/CompletableFutureHelper.kt:31]: Codacy found an issue: [UndocumentedPublicFunction - [runOnFXThread] at /src/src/main/kotlin/insulator/lib/helpers/CompletableFutureHelper.kt:31:49](https://app.codacy.com/manual/darka91/Insulator/pullRequest?prid=5960458)

File: src/main/kotlin/insulator/lib/kafka/model/ConsumerGroup.kt [hunk 14]
```diff
@@ -0,0 +1,7 @@
+package insulator.lib.kafka.model
+
+import org.apache.kafka.common.ConsumerGroupState
+
+data class ConsumerGroup(val id: String, val state: ConsumerGroupState, val members: List<ConsumerGroupMember>)
+data class ConsumerGroupMember(val clientId: String, val topicPartitions: List<TopicPartitionLag>)
```
Comment (darka91) [inline src/main/kotlin/insulator/lib/kafka/model/ConsumerGroup.kt:6]: Codacy found an issue: [UndocumentedPublicClass - [ConsumerGroupMember] at /src/src/main/kotlin/insulator/lib/kafka/model/ConsumerGroup.kt:6:12](https://app.codacy.com/manual/darka91/Insulator/pullRequest?prid=5960458)

**Comment by darka91** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/darka91/Insulator/pullRequest?prid=5960458&bid=19543099)

File: src/test/kotlin/insulator/lib/kafka/AdminApiTest.kt [hunk 15]
```diff
@@ -115,4 +123,37 @@ class AdminApiTest : FunSpec({
         // asssert
         res.get() shouldBeRight {}
     }
+
+    test("describeConsumerGroup happy path") {
+        // arrange
+        val topicName = "topic1"
+        val (consumerGroupOffset, partitionEndOffset) = Pair(1L, 5L)
+        val consumerGroupDescription = mockk<ConsumerGroupDescription> {
+            every { groupId() } returns "consumerGroup"
```
Comment (darka91) [inline src/test/kotlin/insulator/lib/kafka/AdminApiTest.kt:132]: Codacy found an issue: [StringLiteralDuplication - 3/3 - [<anonymous>] at /src/src/test/kotlin/insulator/lib/kafka/AdminApiTest.kt:92:41](https://app.codacy.com/manual/darka91/Insulator/pullRequest?prid=5960458)

File: src/main/kotlin/insulator/views/main/consumergroup/ListConsumerGroupView.kt (hunk 4 above)
Comment (darka91) [inline src/main/kotlin/insulator/views/main/consumergroup/ListConsumerGroupView.kt:21]: Codacy found an issue: [LabeledExpression - [<anonymous>] at /src/src/main/kotlin/insulator/views/main/consumergroup/ListConsumerGroupView.kt:21:48](https://app.codacy.com/manual/darka91/Insulator/pullRequest?prid=5960458)

**Comment by darka91** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/darka91/Insulator/pullRequest?prid=5960458&bid=19543099)

File: src/test/kotlin/insulator/lib/kafka/AdminApiTest.kt (hunk 15 above)
Comment (darka91) [inline src/test/kotlin/insulator/lib/kafka/AdminApiTest.kt:132]: Codacy found an issue: [StringLiteralDuplication - 3/3 - [<anonymous>] at /src/src/test/kotlin/insulator/lib/kafka/AdminApiTest.kt:92:41](https://app.codacy.com/manual/darka91/Insulator/pullRequest?prid=5960458)

File: src/main/kotlin/insulator/views/main/consumergroup/ListConsumerGroupView.kt (hunk 4 above)
Comment (darka91) [inline src/main/kotlin/insulator/views/main/consumergroup/ListConsumerGroupView.kt:21]: Codacy found an issue: [LabeledExpression - [<anonymous>] at /src/src/main/kotlin/insulator/views/main/consumergroup/ListConsumerGroupView.kt:21:48](https://app.codacy.com/manual/darka91/Insulator/pullRequest?prid=5960458)

**Comment by darka91** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/darka91/Insulator/pullRequest?prid=5960458&bid=19543099)

File: src/main/kotlin/insulator/views/main/consumergroup/ListConsumerGroupView.kt (hunk 4 above)
Comment (darka91) [inline src/main/kotlin/insulator/views/main/consumergroup/ListConsumerGroupView.kt:21]: Codacy found an issue: [LabeledExpression - [<anonymous>] at /src/src/main/kotlin/insulator/views/main/consumergroup/ListConsumerGroupView.kt:21:48](https://app.codacy.com/manual/darka91/Insulator/pullRequest?prid=5960458)

File: src/test/kotlin/insulator/lib/kafka/AdminApiTest.kt (hunk 15 above)
Comment (darka91) [inline src/test/kotlin/insulator/lib/kafka/AdminApiTest.kt:132]: Codacy found an issue: [StringLiteralDuplication - 3/3 - [<anonymous>] at /src/src/test/kotlin/insulator/lib/kafka/AdminApiTest.kt:92:41](https://app.codacy.com/manual/darka91/Insulator/pullRequest?prid=5960458)

**Comment by darka91** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/darka91/Insulator/pullRequest?prid=5960458&bid=19543099)

File: src/main/kotlin/insulator/views/main/consumergroup/ListConsumerGroupView.kt (hunk 4 above)
Comment (darka91) [inline src/main/kotlin/insulator/views/main/consumergroup/ListConsumerGroupView.kt:21]: Codacy found an issue: [LabeledExpression - [<anonymous>] at /src/src/main/kotlin/insulator/views/main/consumergroup/ListConsumerGroupView.kt:21:48](https://app.codacy.com/manual/darka91/Insulator/pullRequest?prid=5960458)

**Comment by darka91** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/darka91/Insulator/pullRequest?prid=5960458&bid=19543099)

File: src/main/kotlin/insulator/views/main/consumergroup/ListConsumerGroupView.kt (hunk 4 above)
Comment (darka91) [inline src/main/kotlin/insulator/views/main/consumergroup/ListConsumerGroupView.kt:21]: Codacy found an issue: [LabeledExpression - [<anonymous>] at /src/src/main/kotlin/insulator/views/main/consumergroup/ListConsumerGroupView.kt:21:48](https://app.codacy.com/manual/darka91/Insulator/pullRequest?prid=5960458)

**Comment by andrea-vinci** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/darka91/Insulator/pullRequest?prid=5960458&bid=19543099)

File: src/main/kotlin/insulator/views/main/consumergroup/ListConsumerGroupView.kt (hunk 4 above)
Comment (andrea-vinci) [inline src/main/kotlin/insulator/views/main/consumergroup/ListConsumerGroupView.kt:21]: Codacy found an issue: [LabeledExpression - [<anonymous>] at /src/src/main/kotlin/insulator/views/main/consumergroup/ListConsumerGroupView.kt:21:48](https://app.codacy.com/manual/darka91/Insulator/pullRequest?prid=5960458)

**Pull request closed without merging**